### ISSUE TITLE
How to avoid partition key sorting when inserting data into a partitioned Iceberg table?

### ISSUE THREAD

**Issue by snfddl** [apache/iceberg#10181]
### Query engine

spark 3.2

### Question

1. create partitioned table
```
create table temp.partition_table
(
    dt string
   ,contents string
)
partitioned by spec (dt)
stored as iceberg;
```

2. Insert data with one partition key value into a partitioned table

```
insert into temp.partition_table
select dt
         ,contents
  from temp.dataset
 where dt = '20240418'
```
 
3. physical plan
```
AppendData org.apache.spark.sql.execution.datasources.v2.DataSourceV2Strategy$$Lambda$3485/49797107@3d2cac62, IcebergWrite(table=spark_catalog.temp.partition_table, format=PARQUET)
+- AdaptiveSparkPlan isFinalPlan=false
   +- Sort [dt#107 ASC NULLS FIRST], false, 0   <<<<<<<<<
      +- Exchange hashpartitioning(dt#107, 200), REPARTITION_BY_NUM, [plan_id=175]
         +- Project [dt#107, ansi_cast(contents#106 as string) AS ctnt#110]
            +- FileScan parquet temp.dataset

```

In this case, since dt, the partition key column, has only one value, I don't think there is a need to perform sorting using the partition key right before writing. However, it appears that sorting is always performed using the partition key when inserting data into a partitioned iceberg table.
Is there any way to avoid this?

In impala, unnecessary sorting could be avoided by using the /*+noclustered*/ hint when performing the same type of insert into a general parquet-based table, so I expected a similar method to exist in iceberg, but did not find it.

I also tried static partition insert, but the plan was the same.
(insert into temp.partition_table partition(dt='20240406') select contents from temp.dataset where dt='20240406')

**Comment by snfddl** [comment]
insert into temp.partition_table
select '20240418' as dt
         ,contents
  from temp.dataset
 where dt = '20240418'

**Comment by eubnara** [comment]
How about try? I don't know the reason why....
```
insert into temp.partition_table
select dt
         ,contents
  from temp.dataset
 where dt = '20240418' and dt != NULL
```

**Comment by eubnara** [comment]
How about this? I don't know the reason why but if I add `dt != NULL` condition, `SORT` disappered on `explain`.

```
insert into temp.partition_table
select dt
         ,contents
  from temp.dataset
 where dt = '20240418' and dt != NULL
```

**Comment by lrpt** [comment]
Is there any way to avoid this using the dataframe api?

**Comment by eubnara** [comment]
# tl;dr

When you create a new table with DataFrame api, try to set `use-table-distribution-and-ordering=false` then you may avoid sorting.
Or, how about trying to set table properties as follows.

```
write.distribution-mode=none
write.spark.fanout.enabled=true
```
<img width="829" alt="image" src="https://github.com/user-attachments/assets/a074edfb-25ca-4ff0-9993-b0a74af795e8">


But documentation says memory consumption increases when using spark write fanout feature.


# details


- https://github.com/apache/iceberg/issues/5853
- https://github.com/apache/iceberg/pull/8164#issuecomment-1656277285
- https://github.com/apache/iceberg/blob/apache-iceberg-1.6.1/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkWriteOptions.java#L67-L71
- https://github.com/apache/iceberg/blob/apache-iceberg-1.6.1/core/src/main/java/org/apache/iceberg/util/SortOrderUtil.java#L63-L65


I think if a table is partitioned, sorting occurs as default.



- https://iceberg.apache.org/docs/1.6.1/spark-writes/#writing-distribution-modes
- https://iceberg.apache.org/docs/1.6.1/spark-configuration/#write-options
- https://iceberg.apache.org/docs/1.6.1/configuration/#write-properties

> none - This is the previous default for Iceberg.
This mode does not request any shuffles or sort to be performed automatically by Spark. Because no work is done automatically by Spark, the data must be manually sorted by partition value. The data must be sorted either within each spark task, or globally within the entire dataset. A global sort will minimize the number of output files.
A sort can be avoided by using the Spark write fanout property but this will cause all file handles to remain open until each write task has completed.

### FanoutWriter vs. ClusteredWriter

- FanoutDataWriter extends FanoutWriter
- ClusteredDataWriter extends CLusteredWriter

https://github.com/apache/iceberg/blob/apache-iceberg-1.6.1/core/src/main/java/org/apache/iceberg/io/FanoutWriter.java#L29-L38

https://github.com/apache/iceberg/blob/apache-iceberg-1.6.1/core/src/main/java/org/apache/iceberg/io/ClusteredWriter.java#L33-L41





cf. How about creating `dt` column as `timestamp` or `date` type and create partition using `year()`, `month()`, `day()` or `hour()` and get benefit from iceberg's hidden partioning feature?
If you use it, you don't have to specify partition explicitly when you query.

- https://iceberg.apache.org/spec/#partition-transforms

**Comment by eubnara** [comment]
```
create table temp.partition_table
(
    dt string
   ,contents string
)
partitioned by spec (dt)
stored as iceberg;
```
And this syntax may be impala sql, not sparksql. Is it correct?

`spec` keyword may not be partition transform.

**Comment by lrpt** [comment]
From my side solved with use-table-distribution-and-ordering=false.
Thank you very much @eubnara

**Comment by jiqiujia** [comment]
> From my side solved with use-table-distribution-and-ordering=false. Thank you very much [@eubnara](https://github.com/eubnara)

@lrpt Are there any documentation illustrating this config? I could not find one

**Comment by lzwang2** [comment]
> From my side solved with use-table-distribution-and-ordering=false. Thank you very much [@eubnara](https://github.com/eubnara)

I'm saved!
